src/plots/cartesian/select.js

```diff
@@ -0,0 +1,102 @@
+*/
+
+
+'use strict';
+var polygon = require('../../lib/polygon');
+var axes = require('./axes');
+var filteredPolygon = polygon.filter;
+var polygonTester = polygon.tester;
+var BENDPX = 1.5; // max pixels off straight before a line counts as bent
+
+function getAxId(ax) { return ax._id; }
+
+module.exports = function prepSelect(e, startX, startY, dragOptions, mode) {
+    var plot = dragOptions.plotinfo.plot,
+        dragBBox = dragOptions.element.getBoundingClientRect(),
+        x0 = startX - dragBBox.left,
+        y0 = startY - dragBBox.top,
+        x1 = x0,
+        y1 = y0,
+        path0 = 'M' + x0 + ',' + y0,
+        pw = dragOptions.xaxes[0]._length,
+        ph = dragOptions.yaxes[0]._length,
+        xAxisIds = dragOptions.xaxes.map(getAxId),
+        yAxisIds = dragOptions.yaxes.map(getAxId);
+
+    if(mode === 'lasso') {
+        var pts = filteredPolygon([[x0, y0]], BENDPX);
+    }
+
+    var outlines = plot.selectAll('path.select-outline').data([1,2]);
+
+    outlines.enter()
+        .append('path')
+        .attr('class', function(d) { return 'select-outline select-outline-' + d; })
+        .attr('d', path0 + 'Z');
+
+    // find the traces to search for selection points
+    var searchTraces = [],
+        gd = dragOptions.gd,
+        i,
+        cd,
+        trace,
+        searchInfo,
+        selection = [];
+    for(i = 0; i < gd.calcdata.length; i++) {
+        cd = gd.calcdata[i];
+        trace = cd[0].trace;
+        if(!trace._module || !trace._module.selectPoints) continue;
+
+        if(xAxisIds.indexOf(trace.xaxis) === -1) continue;
+        if(yAxisIds.indexOf(trace.yaxis) === -1) continue;
+
+        searchTraces.push({
+            selectPoints: trace._module.selectPoints,
+            cd: cd,
+            xaxis: axes.getFromId(gd, trace.xaxis),
+            yaxis: axes.getFromId(gd, trace.yaxis)
+        });
+    }
+
+    dragOptions.moveFn = function(dx0, dy0) {
+        var poly;
+        x1 = Math.max(0, Math.min(pw, dx0 + x0));
+        y1 = Math.max(0, Math.min(ph, dy0 + y0));
+
+        if(mode === 'select') {
+            poly = polygonTester([[x0, y0], [x0, y1], [x1, y1], [x1, y0]]);
+            outlines.attr('d', path0 + 'H' + x1 + 'V' + y1 + 'H' + x0 + 'Z');
+        }
+        else if(mode === 'lasso') {
+            pts.addPt([x1, y1]);
+            poly = polygonTester(pts.filtered);
+            outlines.attr('d', 'M' + pts.filtered.join('L') + 'Z');
+        }
+
+        selection = [];
+        for(i = 0; i < searchTraces.length; i++) {
+            searchInfo = searchTraces[i];
+            [].push.apply(selection, searchInfo.selectPoints(searchInfo, poly));
+        }
+        dragOptions.gd.emit('plotly_selecting', {points: selection});
+    };
+
+    dragOptions.doneFn = function(dragged, numclicks) {
+        if(!dragged && numclicks === 2) dragOptions.doubleclick();
+        else {
+            dragOptions.gd.emit('plotly_selected', {points: selection});
```

Comment:
We could also have a single click (without a drag) clear the selection, and leave doubleclick as it is. Not sure which I prefer, just throwing that out there. If we DO choose doubleclick, you're quite right @etpinard that it shouldn't also autoscale.

And when you don't clear, just so we're on the same page this is what I'm imagining (which is different from @cpsievert 's example):
- on mouseup the selection border will disappear but the points will stay selected
- the next selection will draw a new region, which can only add to the existing selection, it can never cause a point to become deselected
- maybe sometime later we can implement a subtractive selection mode, perhaps with a modifier key